xapian-letor/ranker/ranker.cc

```diff
@@ -22,13 +22,280 @@
+    for (size_t i = 0; i < list_fvecs.size(); ++i) {
+	Xapian::FeatureVector fv = list_fvecs[i];
+	double label = fv.get_label();
+	std::vector<double> fvals = fv.get_fvals();
+	Xapian::docid did = fv.get_did();
+
+	train_file << label << " qid:" <<qid;
+	for (int k = 0; k < fv.get_fcount(); ++k) {
+	    train_file << " " << (k+1) << ":" << fvals[k];
+	}
+	train_file << " #docid=" << did<<endl;
+    }
+}
+
+void
+Xapian::prepare_training_file(const string & db_path, const string & queryfile,
+		      const string & qrel_file, Xapian::doccount msetsize,
+		      const string & filename, const FeatureList & flist)
+{
+    // Set db
+    Xapian::Database letor_db(db_path);
+
+    Xapian::SimpleStopper mystopper(sw, sw + sizeof(sw) / sizeof(sw[0]));
+    Xapian::Stem stemmer("english");
+    Xapian::QueryParser parser;
+    parser.add_prefix("title", "S");
+    parser.add_prefix("subject", "S");
+    parser.set_database(letor_db);
+    parser.set_default_op(Xapian::Query::OP_OR);
+    parser.set_stemmer(stemmer);
+    parser.set_stemming_strategy(Xapian::QueryParser::STEM_SOME);
+    parser.set_stopper(&mystopper);
+
+    qrel = load_relevance(qrel_file);
+
+    ofstream train_file;
+    train_file.open(filename);
+
+    string str1;
+    ifstream myfile1;
+    myfile1.open(queryfile.c_str(), ios::in);
+    if(!myfile1.good()) {
+	throw Xapian::FileNotFoundError("No Query file found. Check path.");
+    }
+
+    int query_count = 0;
+    while (!myfile1.eof()) {           // reading all the queries line by line from the query file
+	getline(myfile1, str1);
+	if (str1.empty()) {
+	    break;
+	}
+	++query_count;
+	vector<string> token;
+	size_t j = 0;
+	while (j < str1.size()) {
+	    size_t i = str1.find_first_not_of(' ', j);
+	    if (i == string::npos) break;
+	    j = str1.find_first_of(' ', i);
+	    token.push_back(str1.substr(i, j - i));
+	}
+	// Query file is in the format: <qid> <query_string>
+	// Therefore, <qid> goes into token[0] and <query_string> to token[1]
+
+	// Exceptions for parse errors
+	if (token.size() != 2) {
+	    throw LetorParseError("Could not parse Query file at line:" + Xapian::Internal::str(query_count));
+	}
+	string qid = token[0];
+	string querystr = token[1];
+	if (querystr.front() != '\'' || querystr.back() != '\'') {
+	    throw LetorParseError("Could not parse query string at line:" + Xapian::Internal::str(query_count));
+	}
+	querystr.erase(0, 1); // erase the first character (') from the front
+	querystr.erase(querystr.size() - 1); // erase the last character (')
+	if (querystr.empty()) {
+	    throw LetorParseError("Empty query string in query file at line:" + Xapian::Internal::str(query_count));
+	}
+
+	string qq = querystr;
+	istringstream iss(querystr);
+	string title = "title:";
+	while (iss) {
+	    string t;
+	    iss >> t;
+	    if (t.empty())
+		break;
+	    string temp;
+	    temp.append(title);
+	    temp.append(t);
+	    temp.append(" ");
+	    temp.append(qq);
+	    qq = temp;
+	}
```

Comment:
Is this attempting to add `title:` before every word in the query string?

If so, it would be much better to just specify the default prefix as being the title and parse `querystr`.  The code here's going to get things horribly wrong in non-trivial cases - for example, a quoted phrase.